build/package/i2b2/sql/totalnum-update/generateUpdateTotalnumScript.py

```diff
@@ -0,0 +1,251 @@
+
+
+
+postgreSQL_script = """
+-- set client_min_messages to 'info';
+
+ /* This script updates totalnum for all modifiers and concepts in the specified metadata table.
+  *
+  * The recursive formula for determining the totalnum value of a modifier is:
+  * 	modifier totalnum = #(set of {subject}s directly referencing the modifier UNION of all sets of {subject}s referencing the children of the modifier )
+  *
+  * 	concept totalnum = #(set of {subject}s directly referencing the concept UNION of all sets of {subject}s referencing the children of the concept)
+  *
+  * Rules:
+  * 1) A modifier is a parent of another modifier if the parent's c_fullname is the suffix of another child c_fullname.
+  * 2) A concept is a parent of a modifier if the parent concept c_fullname is equal to the m_applied_path column of the child.
+  * 3) Like rule 1 but the parent and child are both concepts
+  *
+  * Hence one may see multiple layers of modifiers and concepts
+  *
+  * The script uses a bottom approach to compute this formula
+  */
+
+/**
+ * This code traverses iteratively the different heights of modfiers.
+ * At each iteration the code adds the {subject}s that relate to the modifiers to the set of {subject}s that relate to the parent (concept or modifier).
+ * After having done that totalnum is updated for the parent of the current iteration.
+ */
+CREATE OR REPLACE FUNCTION aggregateModifiersCounts() RETURNS void AS $$
+	DECLARE
+		min_depth int := (SELECT MIN(c_hlevel) FROM {metadata_schema_name}.{metadata_table_name} s WHERE LOWER(s.c_facttablecolumn) = 'modifier_cd');
+		max_depth int := (SELECT MAX(c_hlevel) FROM {metadata_schema_name}.{metadata_table_name} s WHERE LOWER(s.c_facttablecolumn) = 'modifier_cd');
+	BEGIN
+
+		raise info 'before for loop';
+		-- the children height is height and the parent height is height-1.
+		FOR height IN REVERSE max_depth..min_depth
+		LOOP
+			raise info 'we are at height %', height;
+
+			--link the {subject}s of the modifier children to their parent modifers or concept.
+			INSERT INTO {metadata_schema_name}.concept_to_id
+				SELECT DISTINCT cti.identifier, parent.c_fullname, (height-1)
+				FROM {metadata_schema_name}.{metadata_table_name} child, {metadata_schema_name}.concept_to_id cti, {metadata_schema_name}.{metadata_table_name} parent
+				WHERE cti.c_hlevel = height AND child.c_hlevel = height AND LOWER(child.c_facttablecolumn) = 'modifier_cd'
+					AND cti.c_fullname = child.c_fullname AND
+					(
+						( --the parent is a modifier at the height just above.
+							parent.c_hlevel = (height - 1) AND LOWER(parent.c_facttablecolumn) = 'modifier_cd' AND
+							child.c_fullname LIKE (parent.c_fullname || '%') ESCAPE '|'
+							-- Use | as an escape character instead of the backslash which appears in windows paths and breaks sql queries.
+						)
+						OR
+						( --the parent is a concept
+							LOWER(parent.c_facttablecolumn) = 'concept_cd' AND parent.c_fullname LIKE child.m_applied_path ESCAPE '|'
```

Comment:
I think there is something I am missing here because I believe this line would take care of m_applied path containing some `%` operator.
For example if you open a postgresql console you can try the following command:
Example of concept (on the left) and a modifier with trailing %. This will print 'ok' because the condition is met. That is the concept path is part of the modifier's applied path.
```psql
select 'ok' where '\abc\def\correct' LIKE '\abc\def\%' escape '|'
```     
Similarly this will return 'ok'
```psql
select 'ok' where '\abc\def\correct\longer\' LIKE '\abc\def\%' escape '|'
```

Reply:
you're right!